File: algorithms-java/src/firstBadVersion/firstBadVersion.java
// Source : https://leetcode.com/problems/first-bad-version/
// Inspired by : http://www.jiuzhang.com/solutions/first-bad-version/
// Author : Lei Cao
// Date   : 2014-10-05

/**********************************************************************************
 *
 * The code base version is an integer start from 1 to n.
 * One day, someone committed a bad version in the code case, so it caused this version and the following versions are all failed in the unit tests.
 * Find the first bad version.
 *
 * You can call isBadVersion to help you determine which version is the first bad one.
 * The details interface can be found in the code's annotation part.
 *
 **********************************************************************************/

package firstBadVersion;

/* The isBadVersion API is defined in the parent class VersionControl.
      boolean isBadVersion(int version); */
public class firstBadVersion extends VersionControl {
    public int firstBadVersion(int n) {
        if (n < 1) {
            return 0;
        }
        int start = 1;
        int end = n;

        while (start + 1 < end) {
            int mid = start + (end - start) / 2;
            if (super.isBadVersion(mid)) {
                end = mid;
            } else {
                start = mid;
            }
        }

        if (super.isBadVersion(start)) {
            return start;
        }
        if (super.isBadVersion(end)) {
            return end;
        }
        return 0;
    }
}

File: algorithms/cpp/firstBadVersion/FirstBadVersion.cpp
// Source : https://leetcode.com/problems/first-bad-version/
// Author : Hao Chen
// Date   : 2015-10-19

/*************************************************************************************** 
 *
 * You are a product manager and currently leading a team to develop a new product. 
 * Unfortunately, the latest version of your product fails the quality check. Since 
 * each version is developed based on the previous version, all the versions after a 
 * bad version are also bad. 
 * 
 * Suppose you have n versions [1, 2, ..., n] and you want to find out the first bad 
 * one, which causes all the following ones to be bad.
 * 
 * You are given an API bool isBadVersion(version) which will return whether version is 
 * bad. Implement a function to find the first bad version. You should minimize the 
 * number of calls to the API.
 * 
 * Credits:Special thanks to @jianchao.li.fighter for adding this problem and creating 
 * all test cases.
 *               
 ***************************************************************************************/

// Forward declaration of isBadVersion API.
bool isBadVersion(int version);

class Solution {
public:
    //Binary search
    int firstBadVersion(int n) {
        int low=1, high=n;
        while(low <= high) {
            int mid = low + (high - low)/2;
            if (isBadVersion(mid) && !isBadVersion(mid-1)){
                return mid;
            }
            if (isBadVersion(mid)) {
                high = mid - 1;
            }else{
                low = mid + 1;
            }
        }
        return -1;
    }
};
